## Tutorial 3 - Cameron O'Dell COMP-SCI 490 Distributed Artifical Intelligence RST

## TASK I

### a. Read Multiple Files using Python

**In the Tutorial Completion Document, describe your findings.**

Please read and print all file names in the "data" folder.

In [57]:
import os

path = 'C:/Users/14175/Desktop/CS490/Tutorial 3/data'
directory_list = os.listdir(path)
print(directory_list)

['.ipynb_checkpoints', 'coffee.csv', 'Employee.csv', 'Employee.json', 'Employee.xlsx', 'HousePrices.csv', 'iris.csv', 'Tutorial3.txt']


**Description of Task I.a**

I imported my os to use the filepath to the data folder. From there, I created a list from os.lisdir on path. Then I printed the list. There are 8 files in the list.

### b. Read and print the top 10 rows from the "coffee.csv" file using Python

**In the Tutorial Completion Document, describe your findings.**

Please read and print the top 10 rows in the **coffee.csv** file in the "data" folder.

In [58]:
import pandas as pd

coffee_file = path+'/coffee.csv'
coffee_df = pd.read_csv(coffee_file)
coffee_first10 = coffee_df.head(10)
print(coffee_first10)

         Date    Open    High     Low   Close  Volume Currency
0  2000-01-03  122.25  124.00  116.10  116.50    6640      USD
1  2000-01-04  116.25  120.50  115.75  116.25    5492      USD
2  2000-01-05  115.00  121.00  115.00  118.60    6165      USD
3  2000-01-06  119.00  121.40  116.50  116.85    5094      USD
4  2000-01-07  117.25  117.75  113.80  114.15    6855      USD
5  2000-01-10  123.50  126.00  116.70  117.55    7499      USD
6  2000-01-11  115.50  118.25  115.50  117.80    3976      USD
7  2000-01-12  117.80  120.50  116.90  118.95    5184      USD
8  2000-01-13  119.25  120.00  117.50  118.55    3717      USD
9  2000-01-14  117.75  120.25  112.25  112.55   10115      USD


**Description of Task I.b**

I imported the pandas module. From there, I created a file destination for coffee.csv to read csv. Then I utilized the header function to pull the first 10 entries. One thing I noticed is that the data is sorted in ascending order from the first date 2000-01-03.

### c. Read and print the last 10 rows from the "coffee.csv" file using Python

**In the Tutorial Completion Document, describe your findings.**

Please read and print the last 10 rows in the **coffee.csv** file in the "data" folder.


In [59]:
coffee_last10 = coffee_df.tail(10)
print(coffee_last10)

            Date    Open    High     Low   Close  Volume Currency
5736  2022-08-22  215.00  225.60  213.55  224.30   10077      USD
5737  2022-08-23  224.55  231.60  222.80  230.65     242      USD
5738  2022-08-24  234.35  242.35  234.35  242.95      62      USD
5739  2022-08-25  242.65  246.90  241.95  243.40      20      USD
5740  2022-08-26  242.60  245.25  242.45  242.00      18      USD
5741  2022-08-29  240.55  240.55  240.55  240.50       1      USD
5742  2022-08-30  238.75  241.75  238.75  239.15      14      USD
5743  2022-08-31  240.75  240.75  239.65  239.10      12      USD
5744  2022-09-01  238.55  238.55  236.60  236.35       6      USD
5745  2022-09-02  232.50  234.05  228.28  228.42       0      USD


**Description of Task I.c**

For this task, instead of using the header function I used the tail function to pull the last 10 entries of the dataset. Nothing to note other than that the most recent data collected is from 2022-09-02. This could be relevant if you were wanting more recent data to work with.

### Activity-I
**In the Tutorial Completion Document, describe your findings.**

a. Print the part of the given arry. Print only even numbers between 20 and 80, including both 20 and 80

In [60]:
import numpy as np

# Arrange elements from 0 to 100
a = np.arange(0, 101)
print("Full Array is:\n",a, '\n')

# To-do: Print only even numbers between 20 and 80, including both 20 and 80
b = np.arange(20, 81, 2)
print("Array with even numbers between 20 and 80 is:\n",b, '\n')

Full Array is:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100] 

Array with even numbers between 20 and 80 is:
 [20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 52 54 56 58 60 62 64 66
 68 70 72 74 76 78 80] 



**Description of Activity I.a**

For this activity, I imported the NumPy module. Then I created an array 'a' from 0 to 100. After that I created an array 'b' from 20 to 80 including only even integers.

### Activity-II
**In the Tutorial Completion Document, describe your findings.**

a. Wite a NumPy program to determine whether any or all of the elements of a given array are zero or non-zero.

In [61]:
x1 = np.array([1, 2, 3, 4])
x2 = np.array([0, 1, 2, 3])

print("Original x1 array: \n", x1)
print("Original x2 array: \n", x2, '\n')

# To-do: write code to check whether any or all of the elements of a given arrays are zero or non-zero.
print(f"None of the elements of the x1 array is zero: {not 0 in x1}\n")
print(f"None of the elements of the x2 array is zero: {not 0 in x2}")

Original x1 array: 
 [1 2 3 4]
Original x2 array: 
 [0 1 2 3] 

None of the elements of the x1 array is zero: True

None of the elements of the x2 array is zero: False


**Description of Activity II.a**

For this activity, the x1 and x2 arrays were given. I used boolean logic to check if zero was in the arrays.

### Activity-III
import pandas as pd
**In the Tutorial Completion Document, describe your findings.**

a. Print the number of rows and columns in the "iris" dataset.

In [62]:
# Reading the CSV file
iris_df = pd.read_csv(path+'/iris.csv')
iris_rows_len = len(iris_df.axes[0])
iris_cols_len = len(iris_df.axes[1])

# To-do: Print the number of rows and columns in the "iris" dataset.
print(f"Number of Rows: {iris_rows_len}\n")
print(f"Number of Columns: {iris_cols_len}")

Number of Rows: 150

Number of Columns: 5


**Description of Activity III.a**

For this activity, I created a dataframe using pandas for the iris.csv dataset. From there, I checked the number of rows and columns by checking the length of the axes. The data set is 150 rows x 5  columns.

**In the Tutorial Completion Document, describe your findings.**

b. Print the mean, standard deviation, minimum and maximum value of **only** '***petal_length***' column of the "iris" dataset

In [63]:
# To-do: Print the mean, standard deviation, minimum and maximum value of only 'petal_length' column of the 'iris' dataset
petal_mean = iris_df['petal_length'].mean()
petal_std = iris_df['petal_length'].std()
petal_mini = iris_df['petal_length'].min()
petal_maxi = iris_df['petal_length'].max()

print (
    f" --Statistical Values of Petal Length in Iris data--\nPetal mean: {petal_mean}\n"
    f"Petal standard deviation: {petal_std}\n"
    f"Petal minimum: {petal_mini}\nPetal maximum: {petal_maxi}"
)

 --Statistical Values of Petal Length in Iris data--
Petal mean: 3.7586666666666693
Petal standard deviation: 1.7644204199522617
Petal minimum: 1.0
Petal maximum: 6.9


**Description of Activity III.b**

For this activity, I utilized the dataframe I had previously created for Iris dataset. Then I used .mean(), .std(), .min(), .max() operations on 'petal_length' column to yield the desired results. If I were presenting this data I might convert the raw float data results by rounding to two or three digits instead of leaving it as is.

**In the Tutorial Completion Document, describe your findings.**

c. Print rows 10 through 20 of the "iris" dataset.

In [64]:
# To-do: Print rows 10 through 20 of the "iris" dataset.
iris_rows_10thru20 = iris_df.iloc[10:21]
print(iris_rows_10thru20)

    sepal_length  sepal_width  petal_length  petal_width      species
10           5.4          3.7           1.5          0.2  Iris-setosa
11           4.8          3.4           1.6          0.2  Iris-setosa
12           4.8          3.0           1.4          0.1  Iris-setosa
13           4.3          3.0           1.1          0.1  Iris-setosa
14           5.8          4.0           1.2          0.2  Iris-setosa
15           5.7          4.4           1.5          0.4  Iris-setosa
16           5.4          3.9           1.3          0.4  Iris-setosa
17           5.1          3.5           1.4          0.3  Iris-setosa
18           5.7          3.8           1.7          0.3  Iris-setosa
19           5.1          3.8           1.5          0.3  Iris-setosa
20           5.4          3.4           1.7          0.2  Iris-setosa


**Description of Activity III.c**

For this activity, I called rows 10 through 20 by using .iloc[] on the iris dataframe I created. Nothing really notable here except that all the species are 'Iris-setosa' in the given range.

**In the Tutorial Completion Document, describe your findings.**

d. Print rows from the "iris" dataset where the species name is **Iris-setosa** and the **petel length** is greater than 1.6.

In [65]:
# To-do: Print rows where species name is "Iris-setosa " and petel length is greater than 1.6.
iris_setosa_petallength_gr1dot6 = iris_df.loc[(iris_df['petal_length'] > 1.6) & (iris_df['species'] == 'Iris-setosa')]
print(iris_setosa_petallength_gr1dot6)

    sepal_length  sepal_width  petal_length  petal_width      species
5            5.4          3.9           1.7          0.4  Iris-setosa
18           5.7          3.8           1.7          0.3  Iris-setosa
20           5.4          3.4           1.7          0.2  Iris-setosa
23           5.1          3.3           1.7          0.5  Iris-setosa
24           4.8          3.4           1.9          0.2  Iris-setosa
44           5.1          3.8           1.9          0.4  Iris-setosa


**Description of Activity III.d**

For this activity, I used .loc on the iris dataframe to return data that fit specific parameters. In this case, if the petal_length was greater than 1.6 and if the species was 'Iris-setosa'. This criteria was only applicable to six entries in the dataset.

## TASK II
### Read the dataset

**In the Tutorial Completion Document, describe your findings.**

a. Read the "HousePrices" dataset from "data" folder

In [66]:
# To-do: Write code to read the "HousePrices" dataset.

houseprices_df = pd.read_csv(path+'/HousePrices.csv', index_col= [0])
houseprices_rows_len = len(houseprices_df.axes[0])
houseprices_cols_len = len(houseprices_df.axes[1])

print(  
    f"Number of Rows: {houseprices_rows_len}\n"
    f"Number of Columns: {houseprices_cols_len}"
)

houseprices_first10 = houseprices_df.head(10)
print(houseprices_first10)

Number of Rows: 546
Number of Columns: 12
      price  lotsize  bedrooms  bathrooms  stories driveway recreation  \
1   42000.0     5850         3          1        2      yes         no   
2   38500.0     4000         2          1        1      yes         no   
3   49500.0     3060         3          1        1      yes         no   
4   60500.0     6650         3          1        2      yes        yes   
5   61000.0     6360         2          1        1      yes         no   
6   66000.0     4160         3          1        1      yes        yes   
7   66000.0     3880         3          2        2      yes         no   
8   69000.0     4160         3          1        3      yes         no   
9   83800.0     4800         3          1        1      yes        yes   
10  88500.0     5500         3          2        4      yes        yes   

   fullbase gasheat aircon  garage prefer  
1       yes      no     no       1     no  
2        no      no     no       0     no  
3        no

**Description of Task II.a**

For this task, I created a dataframe from HousePrices.csv. I added in 'index_col = [0]' to the read command to prevent an 'unamed' index column. I checked the number of rows and columns to better understand the shape of the data. I also called the first 10 rows of the data to get a sample of the data.

### Basic descriptive statistics

**In the Tutorial Completion Document, describe your findings.**

b. Print the basic descriptive statistics like mean, std, etc. for numeric data from the "HousePrices" dataset

In [67]:
# To-do: Write code to print the basic descriptive statistics like mean, std, etc. for numeric data from the "HousePrices" dataset.
# Do not include "Unnamed: 0"
houseprices_statsdescribe = houseprices_df.describe()
print(houseprices_statsdescribe)

print("To-do: \n")

               price       lotsize    bedrooms   bathrooms     stories  \
count     546.000000    546.000000  546.000000  546.000000  546.000000   
mean    68121.597070   5150.265568    2.965201    1.285714    1.807692   
std     26702.670926   2168.158725    0.737388    0.502158    0.868203   
min     25000.000000   1650.000000    1.000000    1.000000    1.000000   
25%     49125.000000   3600.000000    2.000000    1.000000    1.000000   
50%     62000.000000   4600.000000    3.000000    1.000000    2.000000   
75%     82000.000000   6360.000000    3.000000    2.000000    2.000000   
max    190000.000000  16200.000000    6.000000    4.000000    4.000000   

           garage  
count  546.000000  
mean     0.692308  
std      0.861307  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000  
To-do: 



**Description of Task II.b**

For this task, I utilized .describe() from pandas to generate  the mean, std, min, max and quartiles for the numeric data. This tool is very useful as it saves a lot of time calling each function manually. Also worth noting the majority of the homes in the data do not have a garage.

**In the Tutorial Completion Document, describe your findings.**

c. When there are three bedrooms and two bathrooms, what is the average price of a home?

In [68]:

# To-do: Write code to filter the homes where there are three bedrooms and two bathrooms.

houseprices_3bed_2bath = houseprices_df.loc[(houseprices_df['bedrooms'] == 3) & (houseprices_df['bathrooms'] == 2)] 

# To-do: Write code to find the price average from the above subset of data.
houseprices_3bed_2bath_avgprice = houseprices_3bed_2bath['price'].mean()
rounded_3bed_2bath_avgprice = round(houseprices_3bed_2bath_avgprice, 2)
print(f'The average price of a three bedroom, two bathroom home is: $ {rounded_3bed_2bath_avgprice}')

The average price of a three bedroom, two bathroom home is: $ 91136.69


**Description of Task II.c**

For this task, I created a new data frame from houseprices with only homes that have both three bedrooms and two bathrooms. Then I found the mean price of said homes using .mean(). I don't have much to say except there is no way this is the mean price for a home in Kansas City that has three bedrooms and two baths, its too low!

**In the Tutorial Completion Document, describe your findings.**

d. Print the number of houses with a "garage" count of 0 or 1 or 2 or 3


In [69]:
# To-do: Write code to print the number of houses with a "garage" count of 0 or 1 or 2 or 3

houseprices_garage_0to3 = houseprices_df.loc[houseprices_df['garage'] <= 3]
print(houseprices_garage_0to3)

garage_maxi = houseprices_df['garage'].max()
print(f'Maximum garage size is :{garage_maxi}')

        price  lotsize  bedrooms  bathrooms  stories driveway recreation  \
1     42000.0     5850         3          1        2      yes         no   
2     38500.0     4000         2          1        1      yes         no   
3     49500.0     3060         3          1        1      yes         no   
4     60500.0     6650         3          1        2      yes        yes   
5     61000.0     6360         2          1        1      yes         no   
..        ...      ...       ...        ...      ...      ...        ...   
542   91500.0     4800         3          2        4      yes        yes   
543   94000.0     6000         3          2        4      yes         no   
544  103000.0     6000         3          2        4      yes        yes   
545  105000.0     6000         3          2        2      yes        yes   
546  105000.0     6000         3          1        2      yes         no   

    fullbase gasheat aircon  garage prefer  
1        yes      no     no       1     no

**Description of Task II.d**

I created a  dataframe  from all homes with less than or equal to three car carage size (can't have a negative car garage).

I noticed this was the same shape as the dataframe created from the original csv (546 rows by 13 columms). This led me to conclude that all homes in the data have a maximum size of three car garage. I ran a .max() operation on the 'garage' column to confirm this.